syntax='proto3';

// TODO: 1. verify whether the FindReply works as intended: the FindReply object can contain 2 or 3 fields as indicated in the assignment description
// TODO: 2. possibly optimize the classes structure: (after guaranteeing minimal functionality)

// the register service
service Registry {
  rpc register (RegisterRequest) returns (RegisterReply); // called ONLY by Node objects
  rpc deregister (DeregisterRequest) returns (DeregisterReply); // called ONLY by Node objects
  rpc populate_finger_table(PopulateRequest) returns (PopulateReply); // called ONLY by Node objects
  rpc get_chord_info (InfoRequest) returns (InfoReply); // called ONLY BY CLIENT
}
// classes for the register functionality
message RegisterRequest {
  string address = 1;
}

message RegisterReply {
  uint32  id = 1;
  oneof message {
    string error_message = 2;
    uint32 m = 3;
  }
}

// classes for the de register functionality
message DeregisterRequest {
  uint32 address = 1;
}

message DeregisterReply {
  bool result = 1;
  string message = 2;
}

// classes for for populate finger table functionality

message PopulateRequest {
  uint32 node_id = 1;
}

message PopulateReply {
  FingerTableEntry predecessor = 1;
  repeated FingerTableEntry finger_table = 2;
}

message FingerTableEntry {
  uint32 node_id = 1;
  string address = 2;
}

// classes for the get_info_chord functionality


message InfoRequest {}

message InfoReply {
  map<uint32, string> nodes = 1;
}


////////////////////////////////////////////////////////////////////////////////////////////////////////////

// the node service
service Node {
  // the same classes used by the Registry are used by the Node as they have the same fields
  //
  rpc get_finger_table (InfoRequest) returns (InfoReply);
  rpc save_key (SaveRequest) returns (SaveReply);
  rpc remove_key(RemoveRequest) returns (RemoveReply);
  rpc find_key(FindRequest) returns (FindReply);
  rpc quit(QuitRequest) returns (QuitReply);
  rpc notify_predecessor(NotificationRequest) returns (NotificationReply);
  rpc notify_successor(NotificationRequest) returns (NotificationReply)
}

// classes for the save functionality
message SaveRequest {
  string key = 1;
  string text = 2;
}

message SaveReply {
  bool result = 1;
  oneof message {
    uint32 node_id = 2;
    string error_message = 3;
  }
}

// classes for the remove functionality
message RemoveRequest {
  string key = 1;
}

message RemoveReply {
  bool result = 1;
  oneof message {
    uint32 node_id = 2;
    string error_message = 3;
  }
}

// classes for the find functionality

message FindRequest {
  string key = 1;
}


message FindReply {
  bool result = 1;
  oneof message {
    FingerTableEntry node = 2;
    string error_message = 3;
  }
}

// classes for the quit functionality
message QuitRequest {}

message QuitReply {
  string message = 1;
}


message NotificationRequest {
  uint32 sender_id = 1; // the id of the node sending the notification: mainly the node leaving the chord
  uint32 new_id = 2; // either the id of the leaving node's predecessor or successor depending on the recipient node
}

message NotificationReply {}